### ISSUE TITLE
Add Debug Assertion to SecondaryMap::insert catching people trying to insert default keys

### ISSUE THREAD

**Issue by Robert42** [orlp/slotmap#60]
### Problem

I had a bug in my code which resulted in the default key being inserted to a SecondaryMap.

My unittests paniced with the cryptic message:

```
memory allocation of 34359738368 bytes failed
Aborted (core dumped)
```

I am completely aware why this code should not succeed.
But the time to find the bug could have been reduced by if the panic message was more descriptive.

### Suggestion

Add a debug assertion triggering whenever someone by accident passes a default key.

### Minimal Example

I used version 1.0.2

```
#[macro_use]
extern crate slotmap;

use slotmap::SecondaryMap;

new_key_type! {
    pub struct Key;
}

fn main() {
    let mut secondary_map = SecondaryMap::default();
    secondary_map.insert(Key::default(), 42);
}
```

**Comment by orlp** [comment]
For a moment I was worried when I saw 'core dumped' that this was a safety issue, but luckily the 'only' thing that happens is a way too large allocation.

Of course this shouldn't happen, the null key is always considered removed from any slot map. I don't know why I forgot to check for null keys when inserting into a secondary map. I will fix this so that inserting into a null key silently fails (like inserting any removed key can silently fail).

**Comment by Robert42** [comment]
Elegant solution.

It may be a good idea to add a feature allowing users to have debug assertions to catch their own bugs more quickly.

**Comment by orlp** [comment]
Well, I don't think it's necessarily an error to insert into a secondary map with a null key, the fact that this is allowed and silently does nothing allows cleaner implementations sometimes without having to check for null or removed keys explicitly.

E.g. when keeping track of a vertex covering of a tree:

    // Covered is a SecondaryMap<NodeKey, ()>.
    covered.insert(node.parent, ());
    covered.insert(node.left_child, ());
    covered.insert(node.right_child, ());

This is a lot easier than first checking whether the left/right child and parent are valid and exist. And adding a feature flag to enable debug assertions for otherwise perfectly reasonable code (even when behind a feature) might falsely trigger on perfectly fine code.

You can always add your own debug asserts checking for null keys if you suspect they might be an issue in a scenario.

**Comment by orlp** [comment]
Fixed in https://github.com/orlp/slotmap/commit/2d49bbd4fdb6c329c49ba2ac13a929198760364f available in release 1.0.3.

**Comment by Robert42** [comment]
Nice! :)
